components/script/dom/transformstreamdefaultcontroller.rs

```diff
@@ -0,0 +1,458 @@
+        can_gc: CanGc,
+    ) -> DomRoot<TransformStreamDefaultController> {
+        reflect_dom_object(
+            Box::new(TransformStreamDefaultController::new_inherited(transformer)),
+            global,
+            can_gc,
+        )
+    }
+
+    /// Setting the JS object after the heap has settled down.
+    pub(crate) fn set_transform_obj(&self, this_object: SafeHandleObject) {
+        self.transform_obj.set(*this_object);
+    }
+
+    pub(crate) fn set_stream(&self, stream: &TransformStream) {
+        self.stream.set(Some(stream));
+    }
+
+    pub(crate) fn get_finish_promise(&self) -> Option<Rc<Promise>> {
+        self.finish_promise.borrow().clone()
+    }
+
+    pub(crate) fn set_finish_promise(&self, promise: Rc<Promise>) {
+        *self.finish_promise.borrow_mut() = Some(promise);
+    }
+
+    /// <https://streams.spec.whatwg.org/#transform-stream-default-controller-perform-transform>
+    pub(crate) fn transform_stream_default_controller_perform_transform(
+        &self,
+        cx: SafeJSContext,
+        global: &GlobalScope,
+        chunk: SafeHandleValue,
+        can_gc: CanGc,
+    ) -> Fallible<Rc<Promise>> {
+        // Let transformPromise be the result of performing controller.[[transformAlgorithm]], passing chunk.
+        let transform_promise = self.perform_transform(cx, global, chunk, can_gc)?;
+
+        // Return the result of reacting to transformPromise with the following rejection steps given the argument r:
+        rooted!(in(*cx) let mut reject_handler = Some(TransformTransformPromiseRejection {
+            controller: Dom::from_ref(self),
+        }));
+
+        let handler = PromiseNativeHandler::new(
+            global,
+            None,
+            reject_handler.take().map(|h| Box::new(h) as Box<_>),
+            can_gc,
+        );
+        let realm = enter_realm(global);
+        let comp = InRealm::Entered(&realm);
+        transform_promise.append_native_handler(&handler, comp, can_gc);
+
+        Ok(transform_promise)
+    }
+
+    #[allow(unsafe_code)]
+    pub(crate) fn perform_transform(
+        &self,
+        cx: SafeJSContext,
+        global: &GlobalScope,
+        chunk: SafeHandleValue,
+        can_gc: CanGc,
+    ) -> Fallible<Rc<Promise>> {
+        // If transformerDict["transform"] exists, set transformAlgorithm to an algorithm which
+        // takes an argument chunk and returns the result of invoking transformerDict["transform"] with argument list
+        // « chunk, controller » and callback this value transformer.
+        let algo = self.transform.borrow().clone();
+        let result = if let Some(transform) = algo {
+            rooted!(in(*cx) let mut result_object = ptr::null_mut::<JSObject>());
+            rooted!(in(*cx) let mut result: JSVal);
+            rooted!(in(*cx) let this_object = self.transform_obj.get());
+            let call_result = transform.Call_(
+                &this_object.handle(),
+                chunk,
```

Comment:
```assert_equals: desiredSize should still be 0 expected 0 but got 1```


Here you need to pass `argument list « chunk, controller »`, and below you need to pass not `self`. but `result.handle_mut`, otherwise `is_promise` is always false.

Reply:
Ah yes, sorry, you need to use the promise in `Ok(promise) = call_result`, as opposed to using the rooted `result` which is always left as undefined.